src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -0,0 +1,3302 @@
+	execution->tupleDescriptor = tupleDescriptor;
+	execution->tupleStore = tupleStore;
+
+	execution->workerList = NIL;
+	execution->sessionList = NIL;
+	execution->targetPoolSize = targetPoolSize;
+
+	execution->totalTaskCount = list_length(taskList);
+	execution->unfinishedTaskCount = list_length(taskList);
+	execution->rowsProcessed = 0;
+
+	execution->raiseInterrupts = true;
+
+	execution->connectionSetChanged = false;
+	execution->waitFlagsChanged = false;
+
+	return execution;
+}
+
+
+/*
+ * StartDistributedExecution sets up the coordinated transaction and 2PC for
+ * the execution whenever necessary. It also keeps track of parallel relation
+ * accesses to enforce restrictions that arise due to foreign keys to reference
+ * tables.
+ */
+void
+StartDistributedExecution(DistributedExecution *execution)
+{
+	List *taskList = execution->tasksToExecute;
+
+	if (MultiShardCommitProtocol != COMMIT_PROTOCOL_BARE)
+	{
+		if (DistributedExecutionRequiresRollback(execution))
+		{
+			BeginOrContinueCoordinatedTransaction();
+
+			if (TaskListRequires2PC(taskList))
+			{
+				/*
+				 * Although using two phase commit protocol is an independent decision than
+				 * failing on any error, we prefer to couple them. Our motivation is that
+				 * the failures are rare, and we prefer to avoid marking placements invalid
+				 * in case of failures.
+				 */
+				CoordinatedTransactionUse2PC();
+
+				execution->errorOnAnyFailure = true;
+			}
+			else if (MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC &&
+					 list_length(taskList) > 1 &&
+					 DistributedExecutionModifiesDatabase(execution))
+			{
+				/*
+				 * Even if we're not using 2PC, we prefer to error out
+				 * on any failures during multi shard modifications/DDLs.
+				 */
+				execution->errorOnAnyFailure = true;
+			}
+		}
+	}
+	else
+	{
+		/*
+		 * We prefer to error on any failures for CREATE INDEX
+		 * CONCURRENTLY or VACUUM//VACUUM ANALYZE (e.g., COMMIT_PROTOCOL_BARE).
+		 */
+		execution->errorOnAnyFailure = true;
+	}
+
+	execution->isTransaction = InCoordinatedTransaction();
+
+	/*
+	 * We should not record parallel access if the target pool size is less than 2.
+	 * The reason is that we define parallel access as at least two connections
+	 * accessing established to worker node.
+	 *
+	 * It is not ideal to have this check here, it'd have been better if we simply passed
+	 * DistributedExecution directly to the RecordParallelAccess*() function. However,
+	 * since we have two other executors that rely on the function, we had to only pass
+	 * the tasklist to have a common API.
+	 */
+	if (execution->targetPoolSize > 1)
+	{
+		RecordParallelRelationAccessForTaskList(taskList);
+	}
+
+	/* prevent unsafe concurrent modifications */
+	AcquireExecutorShardLocks(execution);
+}
+
+
+/*
+ *  DistributedExecutionModifiesDatabase returns true if the execution modifies the data
+ *  or the schema.
+ */
+static bool
+DistributedExecutionModifiesDatabase(DistributedExecution *execution)
```

Comment:
It might be nice to move these utility functions to the bottom of the file, or a separate file where applicable.

Reply:
I'm not sure if we're still following the conventions, but didn't we have a convention of defining a function right after its first use?